File: ReplayTask.m
#import "ReplayTask.h"
#import "TaskProxy.h"
#import "SchedulerMedusa.h"
#import "TaskScheduler.h"


static inline FileNode * FileNodeFromPath(FileNode *fileTreeRoot, NSString *path, TaskProxy* producer, bool isExclusiveInput)
{
	NSString *lowercasePath = [path lowercaseString];
	const char *posixPath = [lowercasePath fileSystemRepresentation];
	FileNode *outNode = FindOrInsertFileNodeForPath(fileTreeRoot, posixPath);
	//Input nodes don't have a producer.
	//important not to reset to NULL to not override what might have been set already
	if(producer != nil)
	{
		if(outNode->producer != NULL)
		{
			posixPath = [path fileSystemRepresentation];
			fprintf(gLogErr, "error: invalid playlist for concurrent execution.\n"
				"The output path: \"%s\"\n"
				"is specified as a product of two or more actions.\n"
				"See \"replay --help\" for more information about concurrent execution constraints.\n", posixPath);
			safe_exit(EXIT_FAILURE);
		}
		outNode->producer = (__bridge void *)producer;
	}
	else
	{//it is a consumer's request
		if(isExclusiveInput)
			outNode->isExclusiveInput = 1;
		
		if((outNode->isExclusiveInput != 0) && (outNode->hasConsumer != 0))
		{//this input is marked as exclusive and now we are adding a second consumer. This is not allowed
			posixPath = [path fileSystemRepresentation];
			fprintf(gLogErr, "error: invalid playlist for concurrent execution.\n"
				"The path: \"%s\"\n"
				"is specified as an exclusive input for one action (like move or delete) but\n"
				"there is more than one action consuming it.\n"
				"See \"replay --help\" for more information about exclusive inputs.\n", posixPath);
			safe_exit(EXIT_FAILURE);
		}

		outNode->hasConsumer = 1;
	}

	return outNode;
}

NSArray<TaskProxy *> *
TasksFromStep(NSDictionary *replayStep, ReplayContext *context)
{
	if(context->stopOnError && (context->lastError.error != nil))
		return nil;

	__block NSMutableArray<TaskProxy *> *tasksFromStep = [NSMutableArray arrayWithCapacity:0];

	HandleActionStep(replayStep, context,
		^( __nullable dispatch_block_t action,
			NSArray<NSString*> * __nullable inputs,
			NSArray<NSString*> * __nullable exclusiveInputs,
			NSArray<NSString*> * __nullable outputs)
		{
			if(action == nil)
				return;

			TaskProxy *oneTask = [[TaskProxy alloc] initWithTask:action];
			oneTask.stepDescription = replayStep;
			[tasksFromStep addObject:oneTask];

			NSUInteger regularInputCount = 0;
			if(inputs != nil)
				regularInputCount = inputs.count;

			NSUInteger exclusiveInputCount = 0;
			if(exclusiveInputs != nil)
				exclusiveInputCount = exclusiveInputs.count;

			NSUInteger totalInputCount = regularInputCount + exclusiveInputCount;
			FileNode** inputList = NULL;
			if(totalInputCount > 0)
			{
				inputList = (FileNode**)malloc(sizeof(FileNode*) * totalInputCount);
				oneTask.inputCount = totalInputCount;
				oneTask.inputs = inputList;
			}
			
			NSUInteger inputIndex = 0;
			if(inputs != nil)
			{
				for(NSString *oneInput in inputs)
				{
					inputList[inputIndex] = FileNodeFromPath(context->fileTreeRoot, oneInput, nil, false);
					inputIndex++;
				}
			}
			
			if(exclusiveInputs != nil)
			{
				for(NSString *oneInput in exclusiveInputs)
				{
					inputList[inputIndex] = FileNodeFromPath(context->fileTreeRoot, oneInput, nil, true);
					inputIndex++;
				}
			}
			assert(inputIndex == totalInputCount);

			if(outputs != nil)
			{
				NSUInteger outputCount = outputs.count;
				if(outputCount > 0)
				{
					FileNode** outputList = (FileNode**)malloc(sizeof(FileNode*) * outputCount);
					NSUInteger i = 0;
					for(NSString *oneOutput in outputs)
					{
						outputList[i] = FileNodeFromPath(context->fileTreeRoot, oneOutput, oneTask, false);
						i++;
					}
					oneTask.outputCount = outputCount;
					oneTask.outputs = outputList;
				}
			}
		});

	return tasksFromStep;
}


static inline void
ExecuteTasksWithScheduler(NSArray<TaskProxy*> *allTasks, ReplayContext *context, NSUInteger inputCount, NSUInteger outputCount)
{
	ConnectImplicitProducers(context->fileTreeRoot);

	TaskScheduler *scheduler = [[TaskScheduler alloc] initWithConcurrencyLimit:context->councurrencyLimit];

	//graph root task is created by the scheduler
	//we build the graph by adding children tasks to the root

	ConnectDynamicInputsForScheduler(allTasks, //input list of all raw unconnected medusas
									scheduler.rootTask);

	[scheduler startExecutionAndWait];
}

static inline void
VerifyAllTasksExecuted(NSArray<TaskProxy*> *allTasks)
{
	BOOL atLeastOneNotExecuted = NO;
	for(__unsafe_unretained TaskProxy* oneTask in allTasks)
	{
		if(!oneTask.executed)
		{
			if(atLeastOneNotExecuted == NO)
			{ //the first one we encountered
				fprintf(gLogErr, "error: not all tasks have been executed.\n"
				"Most likely there are circular dependencies in the action tree.\n"
				"See \"replay --help\" for more information about action graph restictions.\n"
				"Not executed tasks:\n"
				);
				atLeastOneNotExecuted = YES;
			}
			[oneTask describeTaskToStdErr];
		}
	}
	
	if(atLeastOneNotExecuted)
	{
		safe_exit(EXIT_FAILURE);
	}
}

void
DispatchTasksConcurrentlyWithDependencyAnalysis(NSArray<NSDictionary*> *playlist, ReplayContext *context)
{
	assert(context->concurrent);
	DeleteFileTree(context->fileTreeRoot);
	context->fileTreeRoot = CreateFileTreeRoot();

    NSMutableArray<TaskProxy*> *taskList = [[NSMutableArray alloc] initWithCapacity:0];
	NSUInteger totalInputCount = 0;
	NSUInteger totalOutputCount = 0;

	Class dictionaryClass = [NSDictionary class];

	for(id oneStep in playlist)
	{
		if([oneStep isKindOfClass:dictionaryClass])
		{
			NSArray<TaskProxy *> *stepTasks = TasksFromStep((NSDictionary *)oneStep, context);
			for(TaskProxy *oneTask in stepTasks)
			{
				[taskList addObject:oneTask];
				totalInputCount += oneTask.inputCount;
				totalOutputCount += oneTask.outputCount;
			}
		}
		else
		{
			fprintf(gLogErr, "error: invalid non-dictionary step in the playlist\n");
		}
	}
	
	// at that point the whole input and output paths tree is already constructed
	// with all explicit producers referred in their respective nodes
	
	ExecuteTasksWithScheduler(taskList, context, totalInputCount, totalOutputCount);

	// Post-execution verification
	// in case of circular dependencies some tasks were not scheduled for execution
	// because their dependencies have not been satisifed (the dependency counter never dropped to 0)
	VerifyAllTasksExecuted(taskList);
}

File: replay/ReplayTask.cpp
#include "ReplayTask.h"
#include "TaskCache.h"
#include "TaskProxy.h"
#include "TaskScheduler.h"
#include "SchedulerMedusa.h"
#include "GlobOverlap.h"
#include "ReplaySignpost.h"
#include <algorithm>
#include <cassert>
#include <memory>


static inline FileNode* FileNodeFromPath(FileNode* fileTreeRoot, const std::string& path,
                                         TaskProxy* producer, bool isExclusiveInput)
{
	std::string lowercasePath = path;
	std::transform(lowercasePath.begin(), lowercasePath.end(), lowercasePath.begin(), ::tolower);
	FileNode* outNode = FindOrInsertFileNodeForPath(fileTreeRoot, lowercasePath.c_str());

	if(producer != nullptr)
	{
		if(outNode->producer != nullptr)
		{
			LogError("error: invalid playlist for concurrent execution.\n"
				"The output path: \"%s\"\n"
				"is specified as a product of two or more actions.\n"
				"See \"replay --help\" for more information about concurrent execution constraints.\n", path.c_str());
			safe_exit(EXIT_FAILURE);
		}
		outNode->producer = producer; // raw TaskProxy* stored as void*
	}
	else
	{
		if(isExclusiveInput)
			outNode->isExclusiveInput = 1;

		if((outNode->isExclusiveInput != 0) && (outNode->hasConsumer != 0))
		{
			LogError("error: invalid playlist for concurrent execution.\n"
				"The path: \"%s\"\n"
				"is specified as an exclusive input for one action (like move or delete) but\n"
				"there is more than one action consuming it.\n"
				"See \"replay --help\" for more information about exclusive inputs.\n", path.c_str());
			safe_exit(EXIT_FAILURE);
		}

		outNode->hasConsumer = 1;
	}

	return outNode;
}


// Builds TaskProxy objects from one action step and appends them to the output collections.
// ownedTasks is the lifetime owner; rawList is the non-owning view used by the scheduler.
static void TasksFromStep(const ActionStep& step, ReplayContext* context,
                          std::vector<std::unique_ptr<TaskProxy>>& ownedTasks,
                          std::vector<TaskProxy*>& rawList)
{
	if(context->stopOnError && context->lastError.hasError())
		return;

	HandleActionStep(step, context,
		[&ownedTasks, &rawList, &step, context](
			std::function<bool()> action,
			std::vector<std::string> inputs,
			std::vector<std::string> mutatingInputs,
			std::vector<std::string> exclusiveInputs,
			std::vector<std::string> outputs,
			ActionCacheInfo cacheInfo)
		{
			if(!action)
				return;

			std::string actionName = step.string_value("action").value_or(std::string());

			std::function<void()> taskBlock;
			CacheSession *session = context->cacheSession;
			if((session != nullptr) && cacheInfo.cacheable)
			{
				// Signature and env text come from the expanded, original-case declaration
				// vectors, before the FileTree makes its lowercased copies below, so the
				// cache key stays independent of dependency-analysis internals.
				std::vector<std::string> signatureEnvNames = context->cacheGlobalEnvNames;
				signatureEnvNames.insert(signatureEnvNames.end(), cacheInfo.envNames.begin(), cacheInfo.envNames.end());
				std::string signature = compute_task_signature(actionName, inputs, exclusiveInputs, mutatingInputs, outputs,
					cacheInfo.extras, signatureEnvNames, context->playlistKey);
				std::string envText = build_cache_env_text(context->cacheGlobalEnvNames, cacheInfo.envNames, context->environment);

				// Owned paths (world_out material): outputs + exclusive + mutating, glob or
				// concrete. Concrete outputs separately for the fast existence pass.
				std::vector<std::string> ownedPaths = outputs;
				ownedPaths.insert(ownedPaths.end(), exclusiveInputs.begin(), exclusiveInputs.end());
				ownedPaths.insert(ownedPaths.end(), mutatingInputs.begin(), mutatingInputs.end());

				std::vector<std::string> concreteOutputs;
				for(const auto& oneOutput : outputs)
				{
					if(!globoverlap::is_glob_pattern(oneOutput))
						concreteOutputs.push_back(oneOutput);
				}

				TaskCacheRecord *record = session->make_record(std::move(signature), actionName, inputs,
					std::move(ownedPaths), std::move(concreteOutputs), std::move(envText), cacheInfo.outputsExistenceOnly);
				taskBlock = [session, record, inner = std::move(action)]() {
					session->run_task(record, inner);
				};
			}
			else
			{
				taskBlock = [inner = std::move(action)]() {
					(void)inner();
				};
			}

			auto oneTask = std::make_unique<TaskProxy>(std::move(taskBlock));
			oneTask->stepActionName = actionName;

			TaskProxy* taskPtr = oneTask.get();
			rawList.push_back(taskPtr);
			ownedTasks.push_back(std::move(oneTask));

			// Classify paths: glob patterns go to TaskProxy's glob vectors for
			// overlap-based dependency analysis; concrete paths go into the FileTree
			// for exact node-based dependency tracking.

			std::vector<std::string> globInputList;
			std::vector<std::string> globExclusiveInputList;
			std::vector<std::string> globMutatingInputList;
			std::vector<std::string> concreteMutatingPathList;

			size_t totalInputCount = inputs.size() + exclusiveInputs.size();
			std::unique_ptr<FileNode*, decltype(&free)> inputOwner(
				totalInputCount > 0 ? (FileNode**)malloc(sizeof(FileNode*) * totalInputCount) : nullptr, free);
			FileNode** inputList = inputOwner.get();

			size_t concreteInputIndex = 0;
			for(const auto& oneInput : inputs)
			{
				if(globoverlap::is_glob_pattern(oneInput))
				{
					std::string lowerPath = oneInput;
					std::transform(lowerPath.begin(), lowerPath.end(), lowerPath.begin(), ::tolower);
					globInputList.push_back(std::move(lowerPath));
				}
				else
				{
					inputList[concreteInputIndex] = FileNodeFromPath(context->fileTreeRoot, oneInput, nullptr, false);
					concreteInputIndex++;
				}
			}

			for(const auto& oneInput : exclusiveInputs)
			{
				if(globoverlap::is_glob_pattern(oneInput))
				{
					std::string lowerPath = oneInput;
					std::transform(lowerPath.begin(), lowerPath.end(), lowerPath.begin(), ::tolower);
					globExclusiveInputList.push_back(std::move(lowerPath));
				}
				else
				{
					inputList[concreteInputIndex] = FileNodeFromPath(context->fileTreeRoot, oneInput, nullptr, true);
					concreteInputIndex++;
				}
			}

			for(const auto& oneInput : mutatingInputs)
			{
				if(globoverlap::is_glob_pattern(oneInput))
				{
					std::string lowerPath = oneInput;
					std::transform(lowerPath.begin(), lowerPath.end(), lowerPath.begin(), ::tolower);
					globMutatingInputList.push_back(std::move(lowerPath));
					continue;
				}

				// Concrete mutating path. ConnectGlobDependencies handles producer
				// and consumer chaining uniformly via concreteMutatingPaths and
				// playlist-order Pass B; the FileNode is inserted here for the
				// exclusive-input collision check, the parent-walk in
				// ConnectImplicitProducers (which links a parent dir's producer
				// to the mutator), and to chain a prior playlist producer that
				// only exists as a tree-walk ancestor.
				std::string lowercasePath = oneInput;
				std::transform(lowercasePath.begin(), lowercasePath.end(), lowercasePath.begin(), ::tolower);
				FileNode* node = FindOrInsertFileNodeForPath(context->fileTreeRoot, lowercasePath.c_str());

				if(node->isExclusiveInput != 0)
				{
					LogError("error: invalid playlist for concurrent execution.\n"
						"The path: \"%s\"\n"
						"is specified as a mutating input (e.g. edit) but another action has marked it\n"
						"as an exclusive input (delete or move). These cannot apply to the same path.\n"
						"See \"replay --help\" for more information.\n", oneInput.c_str());
					safe_exit(EXIT_FAILURE);
				}

				// Conditional producer-replacement: only when no prior consumer has registered
				// this path. With prior consumers, the Pass B edge handles ordering instead.
				if(node->hasConsumer == 0)
					node->producer = taskPtr; // raw TaskProxy* as void*

				concreteMutatingPathList.push_back(std::move(lowercasePath));
			}

			if(concreteInputIndex > 0)
			{
				taskPtr->inputCount = concreteInputIndex;
				taskPtr->inputs = inputOwner.release();
			}

			taskPtr->globInputs          = std::move(globInputList);
			taskPtr->globExclusiveInputs = std::move(globExclusiveInputList);
			taskPtr->globMutatingInputs  = std::move(globMutatingInputList);
			taskPtr->concreteMutatingPaths = std::move(concreteMutatingPathList);

			if(!outputs.empty())
			{
				std::unique_ptr<FileNode*, decltype(&free)> outputOwner(
					(FileNode**)malloc(sizeof(FileNode*) * outputs.size()), free);
				FileNode** outputList = outputOwner.get();
				size_t concreteOutputIndex = 0;
				for(const auto& oneOutput : outputs)
				{
					if(globoverlap::is_glob_pattern(oneOutput))
					{
						std::string lowerPath = oneOutput;
						std::transform(lowerPath.begin(), lowerPath.end(), lowerPath.begin(), ::tolower);
						taskPtr->globOutputs.push_back(std::move(lowerPath));
					}
					else
					{
						outputList[concreteOutputIndex] = FileNodeFromPath(context->fileTreeRoot, oneOutput, taskPtr, false);
						concreteOutputIndex++;
					}
				}
				if(concreteOutputIndex > 0)
				{
					taskPtr->outputCount = concreteOutputIndex;
					taskPtr->outputs = outputOwner.release();
				}
			}
		});
}


static inline void
ExecuteTasksWithScheduler(const std::vector<TaskProxy*>& allTasks, ReplayContext* context)
{
	ConnectImplicitProducers(context->fileTreeRoot);

	// Connect glob-based dependencies before dynamic input connection so that
	// glob edges are in place for scheduling.
	ConnectGlobDependencies(allTasks);

	TaskScheduler scheduler(context->councurrencyLimit);
	ConnectDynamicInputsForScheduler(allTasks, scheduler.rootTask());

	REPLAY_SIGNPOST_BEGIN("SchedulerExecution", "task_count=%zu", allTasks.size());
	scheduler.startExecutionAndWait();
	REPLAY_SIGNPOST_END("SchedulerExecution");
}

static inline void
VerifyAllTasksExecuted(const std::vector<TaskProxy*>& allTasks)
{
	bool atLeastOneNotExecuted = false;
	for(TaskProxy* oneTask : allTasks)
	{
		if(!oneTask->executed)
		{
			if(!atLeastOneNotExecuted)
			{
				LogError("error: not all tasks have been executed.\n"
					"Most likely there are circular dependencies in the action tree.\n"
					"See \"replay --help\" for more information about action graph restrictions.\n"
					"Not executed tasks:\n");
				atLeastOneNotExecuted = true;
			}
			oneTask->describeTaskToStdErr();
		}
	}

	if(atLeastOneNotExecuted)
		safe_exit(EXIT_FAILURE);
}

void
DispatchTasksConcurrentlyWithDependencyAnalysis(const std::vector<ActionStep>& playlist, ReplayContext* context)
{
	assert(context->concurrent);
	DeleteFileTree(context->fileTreeRoot);
	context->fileTreeRoot = CreateFileTreeRoot();

	// The cache session owns every TaskCacheRecord and the manifest for the duration
	// of this playlist. One session per --playlist-key: pruning is scoped to the key
	// it processed, so entries belonging to other keys are carried through untouched.
	// MUST be declared before ownedTasks: from stage 4 the task blocks capture record
	// pointers, so the tasks have to be destroyed before the records they point at.
	std::unique_ptr<CacheSession> cacheSession;
	if(context->cacheEnabled && !context->playlistPath.empty())
	{
		cacheSession = std::make_unique<CacheSession>(context->playlistPath, context->playlistKey, context);
		cacheSession->load();
		context->cacheSession = cacheSession.get();
		if(context->verbose)
		{
			LogError("cache: loaded %zu entries from %s\n",
				cacheSession->loaded_entry_count(), cacheSession->manifest_path().c_str());
		}
	}

	std::vector<std::unique_ptr<TaskProxy>> ownedTasks; // lifetime owner
	std::vector<TaskProxy*> taskList;                   // non-owning view for scheduler

	size_t totalInputCount = 0;
	size_t totalOutputCount = 0;

	REPLAY_SIGNPOST_BEGIN("TaskProxyBuild", "playlist_count=%zu", playlist.size());

	for(const auto& step : playlist)
	{
		size_t prevSize = taskList.size();
		TasksFromStep(step, context, ownedTasks, taskList);
		for(size_t i = prevSize; i < taskList.size(); i++)
		{
			totalInputCount  += taskList[i]->inputCount;
			totalOutputCount += taskList[i]->outputCount;
		}
	}

	REPLAY_SIGNPOST_END("TaskProxyBuild");

	if(cacheSession != nullptr)
	{
		// TasksFromStep returns immediately once an error is set under --stop-on-error,
		// including declaration-time errors such as an undefined ${VAR}, so the tail of
		// the playlist may never have been examined. Entries for those steps must not be
		// mistaken for entries of removed steps and pruned.
		bool buildComplete = !(context->stopOnError && context->lastError.hasError());
		cacheSession->set_prune_allowed(buildComplete);
	}

	ExecuteTasksWithScheduler(taskList, context);

	// Note: VerifyAllTasksExecuted calls safe_exit on a dependency cycle, so finalize
	// below does not run in that case. That is safe - every entry is simply carried
	// forward untouched - but it does mean a cycle discards the completed prefix.
	VerifyAllTasksExecuted(taskList);

	// Finalize after the scheduler has drained, and even when lastError is set or
	// --stop-on-error truncated the run: the completed prefix is worth caching, and
	// short-circuited actions report failure so they can never produce a fresh entry.
	// Never write a manifest under --dry-run - nothing actually happened.
	if((cacheSession != nullptr) && !context->dryRun)
	{
		cacheSession->finalize_and_save();
	}
	context->cacheSession = nullptr;

	REPLAY_PRINT_TIMINGS();

	// ownedTasks goes out of scope here -> all TaskProxy destructors run -> free(inputs) / free(outputs) for each task.
}
